crates/wasi-common/src/sys/unix/fdentry_impl.rs

```diff
@@ -56,78 +59,129 @@ pub(crate) unsafe fn determine_type_and_access_rights<Fd: AsRawFd>(
 /// This function is unsafe because it operates on a raw file descriptor.
 pub(crate) unsafe fn determine_type_rights<Fd: AsRawFd>(
     fd: &Fd,
-) -> Result<(
-    wasi::__wasi_filetype_t,
-    wasi::__wasi_rights_t,
-    wasi::__wasi_rights_t,
-)> {
+) -> Result<(types::Filetype, types::Rights, types::Rights)> {
     let (file_type, rights_base, rights_inheriting) = {
+        // Only allow directory operations on directories. Directories can only
+        // yield file descriptors to other directories and files.
+        let rights_dir_base = types::Rights::FD_FDSTAT_SET_FLAGS
```

Comment:
Oh, this one is just a temporary workaround to make `wasmtime` compile. With `wiggle`, `types::Rights::SOME_VALUE` is a constructor of the `types::Rights` struct therefore it's currently impossible to have the composed rights such as `rights_dir_base` as a `const`/`static` unless we resort to `thread_local` or `lazy_static`. The latter is possible, I just didn't want to focus on this at this stage, especially since, I imagine `wiggle` being a path dependency of `wasi-common` in which case we can add more `impl`s to `types::Rights` itself, or perhaps, we could wrap it up in a different transparent type.